### ISSUE TITLE
a solr delete feature

### ISSUE THREAD

**Issue by jrochkind** [traject/traject#166]
Maybe a special key in Traject::Context#clipboard that SolrJsonWriter understands. 

Or maybe an actual new method on Traject::Conext, instead of just a special key in clipboard, clipboard isn't really intended for communicating with writers.

**Comment by jrochkind** [comment]
Hmm, could just be a method on SolrJsonWriter that a traject each_record or other rule can call directly?  Not sure how rules have access to the writer, have to investigate.

**Comment by billdueber** [comment]
Do we have specific use cases? I always generate a list of IDs (or, more likely, have them generated for me) and just shove them at solr in batches. I see three main categories:

* Delete without analysis (e.g, by a set of IDs or a query)
* Delete based on analysis of the record. In this case some of the macros might be useful, but it still feels like a totally different thing. 
* Delete while indexing (basically the same as the previous one, but doing it in one swoop to avoid making two passes through the data). 

If it's really just the first one, I'd think we'd use a separate "deleter" (instead of "indexer"). The reader would be in charge of extracting the unique identifier or query from a list in a file (at the very least we would provide a reader that does  `unique_id = line_from_file.chomp`) and send the delete query to solr in batches. The others still feel like they'd need a different infrastructure than our basic "stick stuff in an accumulator and put that into a hash."

**Comment by jrochkind** [comment]
It's come up in slack. The main concrete use case I can outline is OAI-PMH reader, where some records in the stream are actually 'delete' records. So it's based on analysis, although pretty basic analysis. 

There are some hacky ways to do it without _too_ much code just by writing code, in current traject, of course.  You just `skip!` the record that's really a delete record, and then delete it by some other mechanism -- either right away, or adding to a concurrent-safe list to do later.

**Comment by mejackreed** [comment]
Not sure if this works within what you are doing.. but we need to process a "deletes" file.  Doing something like this: 

https://github.com/sul-dlss/searchworks_traject_indexer/pull/65/files

Also we may want to send "deletes" based off of logic in our main traject config.

Algorithm:

If this record has this thing in marc, send a delete to solr for the id.

**Comment by jrochkind** [comment]
Here's one psuedo-code-ish sketch of a way to implement deletes in the oai use case in existing traject:

```ruby
deleted_ids = Concurrent::Set.new
each_record  do |record, context|
   if is_delete_record?(record)
     context.skip!
     deleted_ids << id_from_delete_record(record)
  end
end

after_processing do
  send_all_deletes_to_solr(deleted_ids)
end
```

That example wouldn't send any deletes until after the entire batch is done, which may or may not be acceptable or desirable in a given use case.

**Comment by jrochkind** [comment]
@mejackread ah, for the entire entirely separate "deletes" file -- of only delete commands?  I wonder if traject needs to handle that at all. Why not just write your own code to process the deletes file and send delete commands to solr?  

You could still use the traject config to find the Solr URL, to DRY your config. But I'm not sure if anything in traject is actually of value here, or if there's any way to make it any easier by traject providing help here. It seems simple enough to just write the ruby in the first place. 

But let me know if I'm missing something or you have another idea!

For "If this record has this thing in marc, send a delete to solr for the id" [presumably _instead_ of an 'add' indexing command] -- that's more like the oai-pmh use case. I think it might make sense for traject to support this somehow, but in the meantime see hacky way to do it manually above what do you think about something like that?

**Comment by mejackreed** [comment]
Yeah I agree with you on this and the OAI way will likely work for us for our marc logic stuff.

Also agree with you about handling separate delete files.. we will tie most of this together soon in the next week so will let you know what we go with.

**Comment by jrochkind** [comment]
SolrJsonWriter has a naive #delete(id) method as of #214. It's not integrated into the indexing pipeline in any way, and also doesn't use batched Solr requests.
